evaluations/ipfs.md

```diff
@@ -0,0 +1,41 @@
+An evaluation of the IPFS paper
+===============================
+
+2021-01-06 Mark Spanbroek
+
+https://ipfs.io/ipfs/QmR7GSQM93Cx5eAg6a6yRzNde1FQv7uL6X1o4k7zrJa3LX/ipfs.draft3.pdf
+
+Goal of this evaluation is to find things to adopt or avoid while designing
+Dagger. It is not meant to be a criticism of IPFS.
+
+#### Pros:
+
++ IPFS is designed by simplifying, evolving, and connecting proven techniques
+  (§3)
++ Consists of a stack of separately described sub-protocols (§3)
++ Uses Coral DSHT to favor data that is nearby, reducing latency of lookup
+  (§2.1.2)
++ Uses proof-of-work in S/Kademlia to discourage Sybil attacks (§2.1.3)
++ Favors self-describing values such as multihash (§3.1) and multiaddr (§3.2.1)
++ BitSwap protocol for exchanging blocks supports multiple strategies (§3.4.2)
++ Uses content addressing (§3.5)
++ The Merkle DAG is simple, yet allows constucting filesystems,
+  key-value stores, databases, messaging system, etc.. (§3.5)
+
+#### Cons:
+
+- Kademlia prefers long-lived nodes, which is not ideal for mobile enviroments
+  (§2.1.1)
+- BitSwap falls just short of introducing a currency with micro payments,
+  necessitating additional work for nodes to find blocks to barter with (§3.4)
+- Merkle DAGs and Paths (§3.5.1) feel like an unnecessary complication at the
```

Comment:
The merkle dag serves as a way of communicating metadata about the chuks, but it's flexible enough to support more sophisticated things. So the merkle dag doesn't actually force anything on the implementation, but it is trivial to implement a virtual unix like file system on top of it for example.

Reply:
Yes, the Merkle DAG is very versatile, and I like that. The point that I was trying to make is that you could probably leave the Merkle DAG out of the base protocol, thereby simplifying it. And then build a layer on top that implements the Merkle DAG.

Reply:
The merkle DAG is mentioned in the paper, but it is indeed a separate component which can be swapped by anything else. The point on bringing it up in the paper (I presume), is to give a full view of an efficient assembly of this components.

Reply:
In paragraph 3.5 they mention that every `IPFSObject` has an array of links (forming the DAG), from which I concluded that it's part of the core of IPFS.

Reply:
Yes, thats misleading as there is no hard requirement for it to be so. But at its core, IPFS is a lego of components (similar to libp2p), that allows swapping different subsystems for other. That said, I like the way the Merkle DAG interacts with the rest of the system.